Added support for Uplift V2 controller

Karuption opened the pull request:
I have just about everything working with as much as the existing code as possible. I would like to extract the logger so that they are easier to implement externally as well as add an enum to convey what exactly is being logged. This would also have a registering function so that it can just proxy the log to all the registered loggers.

phord reviewed (commented):
It's really impossible to review these changes with so many whitespace and style changes mixed in with actual functional ones.  I know my own style in this project is probably inconsistent, but I still don't think I'll have time to merge these upstream as they are.

Karuption commented:
There are some differences that I should fix since readability is worse. You can hide all the whitespace changes when you go to files changed -> Gear icon -> hide whitespace. 

![image](https://user-images.githubusercontent.com/538762/221339628-828c8b59-09a0-486e-b62c-8f81518feaac.png)

phord commented:
That does help and I did that already.  But it doesn't hide line-changes like moving curly-brace to the next line.  I prefer not to take non-functional changes into commits when I can avoid it since it messes around with history so much.  

It's `+361 −63` ignoring space changes, and `+523 −225` without.  So, about 40% smaller that way.  I may try to look closer later, but I'm sure I'll end up filtering a lot of this and committing something altered.

Karuption commented:
I will go through the white space filtered one and cut down the changes tonight and tomorrow. I need to apply your changes from today anyways.

phord commented on Jarvis.ino:
```diff
@@ -42,31 +45,75 @@ void setup() {
+  Jarvis.report();
 }
+
+void onSerialInput() {
```
This belongs in something called SerialServer maybe.  Then the main loop can call `SerialServer.setup()` and `SerialServer.run()` if desired.

Karuption replied:
This was part of serial command interface while I was deving, but was removed as having the software serial and usb serial caused some bugs.

phord commented on src/JarvisDesk.cpp:
```diff
@@ -94,10 +93,8 @@ class JarvisDesk_impl {
           // Press the memory once to try to stop our motion
-          latch_pin(HS0);
-          latch_pin(HS3);
+          press_Memory(30);
```
nit: `snake_case` uses `lower_case`

Karuption replied:
its my C# Pascal case tendencies. fixing.

phord commented on src/JarvisDesk.cpp:
```diff
@@ -107,26 +104,46 @@ class JarvisDesk_impl {
     io_send();
   }
 
+  void press_Memory(int duration = 30) {
```
I figured out that HS0+HS3 isn't the memory button after all.  It's the "wake sequence" to get the controller to start sending height info.  The "Memory" button doesn't actually send anything on my Jarvis desk.  The handset waits for the "preset" button to be pressed, and then it sends the PROGRAM_PRESET command via serial. But the HS0+HS3 is also useful to interrupt the desk from moving, which is all it's needed for here.   Does yours also act this way?

Karuption replied:
memory will stop the desk from working, but it only uses one line. This handset basically never sends anything over UART. When moving for longer periods of times, it will have an almost random sequence that is sent from the handset to the controller. The controller then echos the same values back (even if it is mid "broadcast").

phord replied:
The handset sends an almost random sequence on the UART?  Have you checked to see if it's actually serial data?  Maybe those lines are being used for something else with your controller.  Do you have a logic analyzer that can capture the line signals between the HS and Controller?

Karuption replied:
I do have a cheap logic analyzer I can set up. The only time the Handset sends anything is when the desk is in motion for a longer amount of time (going to a preset or holding up/down). The controller sends back exactly what the handset sent no matter what it is in the middle of sending at the moment. So, I could get a sequence of 1 1 FF 0 FC ( height with a 255 handset echo back in the middle). This language does not appear to be regular so an FA/State Machine is difficult (or maybe even impossible) to model the protocol completely.

I will do some captures this weekend and will try and get as much as I can captured and labeled.

phord commented on src/JarvisDesk.cpp:
```diff
@@ -107,26 +104,46 @@ class JarvisDesk_impl {
+  void reset() {
+    // Press and hold down for 10 seconds
+    latch_pin(HS0);
+    latch_timer.reset(5000);
```
5 seconds?  Or 10?

Karuption replied:
Moving to reset(int duration = 5000), this is the length of time that it takes to clear most errors. but may not be enough to re-level the desk

Karuption replied:
Resolved in my last commit.

phord commented on src/JarvisDesk.cpp:
```diff
@@ -203,33 +220,42 @@ class JarvisDesk_impl {
+    if (preset & 8)
+      latch_pin(HS3);
   }
+#if defined(JCB35N2PA32V2)
```
Oh, is this the 0x01 0x01 protocol I saw before?  Nice.

Karuption replied:
yes, makes it a bit difficult with how simplistic it is

phord replied:
I have tricked my controller into using this mode accidentally. I don't know how. I wonder what your handset does to force it always into this mode. It woulc be nice to see a logic analyzer capture of the signals on startup from your controller.

Karuption replied:
I will add it to the list of things to capture. I would rather use your packet mode (if that is even possible) anyways.

phord commented on src/JarvisDesk.cpp:
```diff
@@ -242,9 +268,11 @@ class JarvisDesk_impl {
-
     h = Util::to_mm(h);
-    if (height == h) return;
+    if (height == h || h < 640 || h > 1300)
```
Why are these out of range?  Can you math here or use a #define?

Karuption replied:
This was the range of quite of the few desks I found. I will add them to a #define in the pinout config.

phord replied:
`to_mm()` already does `h = height * 2.54` when `height < 600`.  So `h > 1300` should only happen if `height > 1300` as well, I think.  I already filtered out "9999" as a bogus value I receive from my desk sometimes.  I wonder what other sentinel values we could see here.

phord commented on src/JarvisDesk.cpp:
```diff
@@ -260,57 +288,70 @@ class JarvisDesk_impl {
-      ARGS = LENGTH + sizeof(argv),   // collecting args
-      CHKSUM,  // waiting for checksum
-      ENDMSG,  // waiting for EOM
+      ARGS = sizeof(argv), // LENGTH + sizeof(argv), // collecting args
```
This will break my other protocol

phord reviewed (commented):
Ok, it wasn't so much to review, I suppose.

I was thinking of having a logging "token" for each log line I want to be able to filter. Then Like `log.println("HS:", "Handset button ", btn, " pressed.");` Then I could have a cmd interface that lets me `enable HS` or `disable HS` to enable and disable associated logging.  

The log module, telnet module, etc. are ones I've developed and used for other projects. I'd like to make them more modular (and in some cases, I have). But I have a tendency to hyperfixate on some piece and overengineer it.  So I try to resist to urges.  :-)

Karuption commented:
> I was thinking of having a logging "token" for each log line I want to be able to filter

I was thinking the same thing. I was thinking of possibly an enum for all tags so you can make a switch in a logger to handle some logs differently. If you have multiple loggers, you could ignore a DEBUG message but also turn it on at any time (sending telnet a 'debug' command for instance.

>   I'd like to make them more modular (and in some cases, I have). But I have a tendency to hyperfixate on some piece and overengineer it. So I try to resist to urges. :-)

I feel that, but that was something that I was thinking about too, I would like to be able to log to a custom MQTT address for better metrics and being able to send commands via topic.

phord commented on src/JarvisDesk.cpp:
```diff
@@ -203,15 +218,15 @@ class JarvisDesk_impl {
   }
 
-  void unlatch()
+  void unlatch() 
```
nit: No spaces at end of line

phord commented on src/JarvisDesk.cpp:
```diff
@@ -203,15 +218,15 @@ class JarvisDesk_impl {
   }
 
-  void latch(unsigned preset)
+  void latch(unsigned preset) 
```
space at eol

Karuption commented on src/JarvisDesk.cpp:
```diff
@@ -147,34 +161,35 @@ class JarvisDesk_impl {
-private:
-  AdafruitIO_Group *jarvis = nullptr;
+  private:
+  AdafruitIO_Group* jarvis = nullptr;
```
I always use multiple lines for declarations and view the * as part of the type vs part of the variable, but will fix and keep this in mind for this  project.

Karuption commented on src/JarvisDesk.cpp:
```diff
@@ -267,50 +289,64 @@ class JarvisDesk_impl {
       // HANDSET
-      UNITS       = 0x0E,  // Set units to cm/inches
-      MEM_MODE    = 0x19,  // Set memory mode
-      COLL_SENS   = 0x1D,  // Set anti-collision sensitivity  (Sent 1x; no repeats)
-      SET_MAX     = 0x21,  // Set max height; Sets max-height to current height
-      SET_MIN     = 0x22,  // Set min height; Sets min-height to current height
-      LIMIT_CLR   = 0x23,  // Clear min/max height
-      PROGMEM_1   = 0x03,  // Set memory position 1 to current height
-      PROGMEM_2   = 0x04,  // Set memory position 2 to current height
-      PROGMEM_3   = 0x25,  // Set memory position 3 to current height
-      PROGMEM_4   = 0x26,  // Set memory position 4 to current height
-      WAKE        = 0x29,  // Poll message (??) sent when desk doesn't respond to BREAK messages
-      CALIBRATE   = 0x91,  // Height calibration (Repeats 2x)
+      UNITS = 0x0E,     // Set units to cm/inches
+      MEM_MODE = 0x19,  // Set memory mode
+      COLL_SENS = 0x1D, // Set anti-collision sensitivity  (Sent 1x; no repeats)
+      SET_MAX = 0x21,   // Set max height; Sets max-height to current height
+      SET_MIN = 0x22,   // Set min height; Sets min-height to current height
+      LIMIT_CLR = 0x23, // Clear min/max height
+      PROGMEM_1 = 0x03, // Set memory position 1 to current height
+      PROGMEM_2 = 0x04, // Set memory position 2 to current height
+      PROGMEM_3 = 0x25, // Set memory position 3 to current height
+      PROGMEM_4 = 0x26, // Set memory position 4 to current height
+      WAKE = 0x29,      // Poll message (??) sent when desk doesn't respond to BREAK messages
+      CALIBRATE = 0x91, // Height calibration (Repeats 2x)
     };
```
Matched my alignment to your orig alignment.

phord commented:
I have a clang-format config defined in my project locally, but I've obviously not used it since my indents are all wonky.  But now I tried it and I see it makes many of the changes you submitted, including ones I objected to.  Pity about some of those.   I suppose that's why I still don't use it regularly.  :-(

I've experimented with refactoring the protocol code out of JarvisDesk.cpp to facilitate incorporating your changes.  Then just as I finished it up, it occurred to me that perhaps the protocols should be blended more. It seems like the original designers were using something like MODBUS where the ADDRESS byte tells us which "protocol" to expect.  It seems more complicated that it's worth, though.  I'll take a stab at merging your protocol into my refactoring and post it for you to take a look.

Thanks for your efforts so far to get this other protocol incorporated.

phord commented:
I've merged your changes into my protocol refactoring here: [phord/jarvis:uplift](https://github.com/phord/Jarvis/tree/uplift).  Can you try it out and see if I missed anything?

I removed `pending_reset` which seemed to be unused, and I dropped the `WiFi::accept` change in TelnetLogger, because my library doesn't support it yet.

Karuption commented:
> I removed `pending_reset` which seemed to be unused

Pending _reset was for basically holding down the down button to clear desk errors (I'm short and the desk is unevenly weighted). It likely could be incorporated to a Down(int duration) and an Up(int duration) method since those would be the only ones that aren't exposed. Also, a Stop() exposed interface would likely be a good idea. Just unlatch and set the moving boolean or hit memory for 30ms. All of these would be an easy addition.

> I dropped the `WiFi::accept` change in TelnetLogger, because my library doesn't support it yet.

I was actually about to test that regression since I have been noticing weird telnet bugs. Your current method is marked deprecated, but that seems to be a fairly new (and less stable) option.

I'll clone the other branch and report back (likely with some logic analyzer stuff).

Karuption commented:
Everything seems to be stable with this branch.

phord commented:
> > I removed `pending_reset` which seemed to be unused
> 
> Pending _reset was for basically holding down the down button to clear desk errors

I understood what it was intended for, but it was never actually read or set to anything after initialization. I was worried I might have lost something in the refactoring, but I didn't find any other mentions of it in your branch, either.  Feel free to add it back in when you need it.

phord commented:
I added separate loggers that can be enabled/disabled.   I don't have a generic registration, but it wouldn't be too hard to add.  Instead of enums, I just wrapped them in different instances.  Take a look at f54829c.

But also be warned - I switched the default pinouts config back to Jarvis/Wemos.
